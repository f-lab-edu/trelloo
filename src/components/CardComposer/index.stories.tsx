import React, { useState } from "react";
import styled from "styled-components";
import CardComposer from "./index";

export default {
  title: "components/forms/CardComposer",
  component: CardComposer,
};

const Template = () => {
  const [isCardInputOpened, setIsCardInputOpened] = useState(false);
  const [isLoading, setLoading] = useState(false);
  const handleAddCard = async () => {
    setLoading(true);
    setTimeout(() => {
      setLoading(false);
    }, 2000);
  };
  const handleCardInputToggle = () => {
    setIsCardInputOpened(!isCardInputOpened);
  };

  return (
    <Container>
      <CardComposer
        isLoading={isLoading}
        isCardInputOpened={isCardInputOpened}
        onCardInputToggle={handleCardInputToggle}
        listId="id"
        onAddCard={handleAddCard}
      />
    </Container>
  );
};

export const Default = Template.bind({});

const Container = styled.div`
  width: 300px;
`;
