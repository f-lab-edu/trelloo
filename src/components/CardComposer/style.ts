import { theme } from "@/styles/theme";
import styled from "styled-components";

export const Container = styled.div`
  //
`;

export const Header = {
  minHeight: "40px",
  backgroundColor: theme.color.cardListGray,
  fontSize: "14px",
};

export const Body = {
  padding: "0 10px 5px 10px",
  width: "100%",
  backgroundColor: theme.color.cardListGray,
};

export const ButtonWrapper = styled.div`
  margin-top: 5px;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

export const TemplateButton = {
  width: 24,
};

export const CardInputContainer = styled.div`
  margin-top: 10px;
`;

export const CardInput = {
  padding: 0,
  height: 65,
  boxShadow: theme.boxShadow.card,
};

export const AddCardButtonContainer = styled.div`
  padding: 5px 0;
  display: flex;
  justify-content: space-between;
`;

export const AddCardButtonWrapper = styled.span`
  display: flex;
  align-items: center;
`;

export const CancelAddCardButton = {
  marginLeft: 5,
  padding: 0,
  fontSize: 20,
  color: theme.color.textPointGray,
};
