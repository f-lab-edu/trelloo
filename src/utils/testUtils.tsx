import { store } from "@/store";
import { GlobalStyle } from "@/styles/GlobalStyle";
import { theme } from "@/styles/theme";
import ModalsProvider from "@components/modals/ModalsProvider";
import { QueryClient, QueryClientProvider, UseQueryResult } from "@tanstack/react-query";
import { render, renderHook, RenderOptions } from "@testing-library/react";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";

const TestProviders = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <GlobalStyle />
          <ModalsProvider>{children}</ModalsProvider>
        </ThemeProvider>
      </Provider>
    </QueryClientProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: TestProviders, ...options });

const createQueryClientWrapper = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient();

  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

export const createMockedQuery = (hook: () => UseQueryResult<any, unknown>) => {
  return renderHook(() => hook(), { wrapper: createQueryClientWrapper });
};

export * from "@testing-library/react";
export { customRender };
