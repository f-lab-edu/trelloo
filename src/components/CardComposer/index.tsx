import { useState } from "react";
import { Input, Spin } from "antd";
import { AddCardRequest } from "@/queries/cards/interface";
import Button from "@components/Button";
import { EllipsisOutlined, PlusOutlined, PicLeftOutlined } from "@ant-design/icons";
import * as S from "./style";

const { TextArea } = Input;

interface Props {
  isLoading: boolean;
  isCardInputOpened: boolean;
  onCardInputToggle: () => void;
  listId: string;
  onAddCard: (params: AddCardRequest) => void;
}

const CardComposer = ({ isCardInputOpened, onCardInputToggle, listId, onAddCard, isLoading }: Props) => {
  const [cardInputValue, setCardInputValue] = useState("");

  const handleAddCard = ({ description, listId }: AddCardRequest) => {
    onAddCard({ description, listId });
    setCardInputValue("");
  };

  return (
    <>
      {isCardInputOpened ? (
        <S.CardInputContainer>
          <Spin spinning={isLoading}>
            <S.Card>
              <TextArea
                value={cardInputValue}
                onChange={(e) => setCardInputValue(e.target.value)}
                placeholder="Enter a title for this card..."
                autoSize
                bordered={false}
              />
            </S.Card>
          </Spin>

          <S.AddCardButtonContainer>
            <S.AddCardButtonWrapper>
              <Button
                appearance={{ type: "blue" }}
                onClick={() => handleAddCard({ description: cardInputValue, listId })}
                isLoading={isLoading}
              >
                Add card
              </Button>
              <S.CloseButton onClick={onCardInputToggle} />
            </S.AddCardButtonWrapper>
            <Button appearance={{ type: "transparent" }} Icon={<EllipsisOutlined />} />
          </S.AddCardButtonContainer>
        </S.CardInputContainer>
      ) : (
        <S.ButtonWrapper>
          <Button appearance={{ type: "transparent" }} Icon={<PlusOutlined />} onClick={onCardInputToggle}>
            Add a card
          </Button>
          <Button appearance={{ type: "transparent" }} Icon={<PicLeftOutlined />} />
        </S.ButtonWrapper>
      )}
    </>
  );
};

export default CardComposer;
