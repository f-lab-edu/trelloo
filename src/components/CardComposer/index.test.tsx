import React, { useState } from "react";
import { customRender, fireEvent, renderHook } from "@utils/testUtils";
import { describe, vi, expect } from "vitest";
import CardComposer from ".";
import "@testing-library/jest-dom";
import { act, cleanup } from "@testing-library/react";

describe("CardComposer 테스트", () => {
  afterEach(cleanup);
  const handleAddCard = vi.fn();

  function useCardState() {
    const [isCardInputOpened, setIsCardInputOpened] = useState(false);

    const handleCardInputToggle = () => {
      setIsCardInputOpened(!isCardInputOpened);
    };
    return { isCardInputOpened, handleCardInputToggle };
  }

  const { result } = renderHook(useCardState);

  const setup = (isCardInputOpened: boolean, handleCardInputToggle: () => void) => {
    const { getByText, getByPlaceholderText, getByLabelText } = customRender(
      <CardComposer
        isLoading={false}
        isCardInputOpened={isCardInputOpened}
        onCardInputToggle={handleCardInputToggle}
        listId="listIs"
        onAddCard={handleAddCard}
      />,
    );

    return { getByText, getByPlaceholderText, getByLabelText };
  };

  it("텍스트 입력 후 카드 생성", async () => {
    const { getByPlaceholderText, getByText } = setup(
      result.current.isCardInputOpened,
      result.current.handleCardInputToggle,
    );
    const addCardButton = getByText("Add card");
    const input = getByPlaceholderText("Enter a title for this card...") as HTMLInputElement;

    act(() => {
      fireEvent.change(input, { target: { value: "text" } });
    });

    expect(input.value).toEqual("text");

    act(() => {
      fireEvent.click(addCardButton);
    });

    expect(handleAddCard).toHaveBeenCalled();
  });
});
