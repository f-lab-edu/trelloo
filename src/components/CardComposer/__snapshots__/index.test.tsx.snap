// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardComposer 테스트 > 스냅샷 테스트 1`] = `
{
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByText": [Function],
}
`;
